text/0045-add-notion-of-protocol.md

```diff
@@ -0,0 +1,144 @@
+  - Up to one URL prefix per protocol in `protocols`.
+  - Those URL prefixes must be in the same order as `protocols`.
+  - Each URL prefix must be unique to the protocol
+  - Meant to be used for protocol auto-detection on server URL.
+  - While registering if present, the number of URL prefix must be less or equal to the number of protocols.
+  - NOTE: URL prefixes (URI scheme + `://`) are prefered over URI schemes to avoid:
+    - matching links without prefixes (e.g. "rtmp.example.com").
+    - adding `://` each time that a auto-detection is done.
+
+Adding to this API these functions:
+- `const char *obs_output_get_protocols(const obs_output_t *output)`: returns protocols supported by the output.
+- `bool obs_output_is_protocol_registered(const char *protocol)`: return true if an output with the protocol is registered.
+- `const char *obs_output_get_prefix_protocol(const char *prefix)`: return the protocol bound to the prefix.
+- `bool obs_enum_output_service_types(const char *protocol, size_t idx, const char **id)`: enumerate all outputs types compatible with the given protocol.
+- `bool obs_enum_output_protocols_prefixes(size_t idx, const char **prefix)`: enumerate all registered URL prefixes.
+- `bool obs_enum_output_protocols(size_t idx, const char **prefix)`: enumerate all registered protocol.
+- `const char *obs_get_output_supported_video_codecs(const char *id)`: return compatible video codecs of the given output id.
+- `const char *obs_get_output_supported_audio_codecs(const char *id)`: return compatible audio codecs of the given output id.
+
+### About protocols
+
+- HLS relies on HTTP/HTTPS and it is not the only existing protocol in this case. This protocol has no unique URL prefixes.
+
+- Because of the deprecation of the protocol, FTL will not have a URL prefix asign to its output to not make it available through custom server in the UI.
+
+## Services API
+Since a streaming service may not accept all the codecs usable by a protocol, adding a way to set supported codecs is required.
+
+Adding to `obs_service_info`:
+- `const char *(*get_protocol)(void *data)`: return the protocol used by the service. RFC 39 will allow multi-protocol services in the future.
+- `const char **(*get_supported_video_codecs)(void *data)`: video codecs supported by the service. Optional, fallback to protocol supported codecs if not set.
+- `const char **(*get_supported_audio_codecs)(void *data)`: audio codecs supported by the service. Optional, fallback to protocol supported codecs if not set.
+
+Adding to this API these functions:
+- `const char *obs_service_get_protocol(const obs_service_t *service)`: return the protocol used by the service.
+- `const char **obs_service_get_supported_video_codecs(const obs_service_t *service)`: return video codecs compatible with the service.
+- `const char **obs_service_get_supported_audio_codecs(const obs_service_t *service)`: return audio codecs compatible with the service.
```

Comment:
@tytan652 Audio/Video codecs aren't know ahead of time for WebRTC. The offerer sends a list of all supported codecs, and the answerer returns the intersection.

The use case I think is important is `OBS -> Browser`. Users will be able to stream directly to a friend with no server/backend required. We can't configure ahead of time what codecs are going to work.

Reply:
If the streaming service does not put a list of compatible codec in the services JSON, it will allow any codec that is set in the output. But the service **must** support al least H264 and AAC or opus.

If it is a "Custom server", then it will allow any codec that is set in the output. But in this scenario the user is by himself.

> The offered sends a list of all supported codecs, and the answerer returns the intersection.

The user only set one encoder (e.g. NVENC H264, AOM AV1… ) in the settings, only codecs supported by the service and the outputs are shown in the settings UI.

The user (auto-wizard) decides which encoder (one codec) it uses, the UI and the output **must** follow user settings.

This is not part of the RFC, this is how OBS Studio actually works with a hard-coded codecs list.
The RFC just plan to replace the list with service supported codecs.

> Could codec determination be made later then service/protocol registration?

It could be done while calling `obs_service_get_supported_`video/audio`_codecs()` in services `get_supported_`video/audio`_codecs`.

But output codecs are fixed.

Reply:
@tytan652 mind unresolving my comment, I don't believe it has been addressed?

The goal of this RFC is `Create a better management of outputs, encoders, and services with an approach focused on protocols.`. If we don't address use cases we know about another RFC + refactor will have to be done anyway. The current design satisfies the goals of the RFC.

`obs_service_get_supported_video` I don't believe would be the right place. The available codecs will change for every connection. The remote host doesn't have a static list of codecs they support.

Reply:
Without (or with) the RFC, the output does not (and will not) negotiate the codec when starting the stream/connection and suddenly change the encoder.
The output uses the encoder it was given and works with it or fails, nothing else.

- The service beside "Custom…" must provide its supported codecs in the JSON if it does not support all the one that the output could support. And the user is unable to choose unsupported codecs in the settings.
- If "Custom…", it falls back to what the output support, the user should know which codec/encoder it should. And if it this not supported while connecting, it fails.